### PR TITLE
dp c_allreduce_sum_fusion op

### PR DESCRIPTION
**Pull request by Baibaifan**
### PR types

Function optimization
### PR changes

OPs
### Describe

7f440991d5724b3098c16eb08277b120

dp c_allreduce_sum_fusion op

### CODE REVIEW

File: paddle/fluid/framework/distributed_strategy.proto [hunk 1]
```diff
@@ -176,6 +176,8 @@ message DistributedStrategy {
   optional bool find_unused_parameters = 28 [ default = false ];
   optional bool tensor_parallel = 29 [ default = false ];
   optional bool without_graph_optimization = 30 [ default = false ];
+  optional bool allreduce_fusion_optimization = 31 [ default = false ];
```
Comment (wangxicoding) [inline paddle/fluid/framework/distributed_strategy.proto]: 之前有fuse_all_reduce_ops、fuse_grad_size_in_MB、fuse_grad_size_in_TFLOPS这几个选项，复用就行

File: paddle/fluid/operators/coalesce_tensor_op.cc [hunk 2]
```diff
@@ -145,21 +145,40 @@ class CoalesceTensorOpKernel : public framework::OpKernel<T> {
     offset = 0;
     std::stringstream ss;
     ss << "alloc_space_for_vars: ";
+#if defined(PADDLE_WITH_ASCEND_CL)
+    auto stream =
+        context.template device_context<paddle::platform::NPUDeviceContext>()
+            .stream();
+    platform::NPUMemsetAsync(
```
Comment (wangxicoding) [inline paddle/fluid/operators/coalesce_tensor_op.cc:153]: why memset, add TODO use set_constant

File: paddle/fluid/operators/coalesce_tensor_op.cc [hunk 3]
```diff
@@ -145,21 +145,40 @@ class CoalesceTensorOpKernel : public framework::OpKernel<T> {
...
+#endif
     for (size_t i = 0; i < out_tensors.size(); ++i) {
       size_t len = static_cast<size_t>(out_tensors[i]->numel());
       auto dim = out_tensors[i]->dims();
       out_tensors[i]
           ->ShareDataWith(fused_tensor->Slice(
               static_cast<int64_t>(offset), static_cast<int64_t>(offset + len)))
           .Resize(dim);
+#if defined(PADDLE_WITH_ASCEND_CL)
+      len = use_align ? len + platform::NPUMinChunkSize() : len;
```
Comment (wangxicoding) [inline paddle/fluid/operators/coalesce_tensor_op.cc]: implement Alignment

File: paddle/fluid/operators/coalesce_tensor_op.cc [hunk 4]
```diff
@@ -191,13 +210,19 @@ class CoalesceTensorOpKernel : public framework::OpKernel<T> {
       ss << "input(" << var_names[i] << ") dim:(" << lod_tensors[i]->dims()
          << ") "
          << " addres:" << lod_tensors[i]->data<void>() << ", ";
+
+#if defined(PADDLE_WITH_ASCEND_CL)
+      *numel += use_align
```
Comment (wangxicoding) [inline paddle/fluid/operators/coalesce_tensor_op.cc]: implement Alignment

File: python/paddle/distributed/fleet/base/distributed_strategy.py [hunk 5]
```diff
@@ -853,6 +853,50 @@ def without_graph_optimization(self, flag):
                 "WARNING: without_graph_optimization should have value of bool type"
             )
 
+    @property
+    def allreduce_fusion_optimization(self):
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/base/distributed_strategy.py]: 复用fuse_all_reduce_ops

File: python/paddle/distributed/fleet/base/distributed_strategy.py [hunk 6]
```diff
@@ -853,6 +853,50 @@ def without_graph_optimization(self, flag):
...
+    def allreduce_fusion_optimization(self, flag):
+        if isinstance(flag, bool):
+            self.strategy.allreduce_fusion_optimization = flag
+        else:
+            print(
+                "WARNING: allreduce_fusion_optimization should have value of bool type"
+            )
+
+    @property
+    def allreduce_fusion_split_num(self):
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/base/distributed_strategy.py]: 复用之前的，或者换个和之前名称类似的

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 7]
```diff
@@ -123,7 +129,11 @@ def minimize_impl(self,
 
     def _transpile_main_program(self, loss):
         self._insert_loss_grad_ops(loss)
-        self._insert_allreduce_ops()
+        if self.allreduce_fusion_optimization and core.is_compiled_with_npu():
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py]: add gpu fuse

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 8]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
                     attrs={'ring_id': ring_id,
                            OP_ROLE_KEY: OpRole.Backward})
                 break
+
+    def _allreduce_fusion_program(self):
+        block = self.main_program.global_block()
+        ring_id = self.global_ring_id
+        record_idx, allreduce_input_vars, allreduce_output_vars = [], [], []
+        use_align = True if core.is_compiled_with_npu() else False
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py]: default True

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 9]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
...
+                            continue
+                        record_idx.append([0, idx])
+                    elif len(allreduce_output_vars[
+                            -1]) == self.allreduce_fusion_split_num:
+                        allreduce_output_vars.append([grad])
+                        allreduce_input_vars.append([param])
+                        if self.allreduce_fusion_split_num == 1:
+                            record_idx.append([idx, idx])
+                            continue
+                        if idx != len(list(enumerate(block.ops))) - 1:
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py]: use var to cache `len(list(enumerate(block.ops)))`

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 10]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
                     attrs={'ring_id': ring_id,
                            OP_ROLE_KEY: OpRole.Backward})
                 break
+
+    def _allreduce_fusion_program(self):
+        block = self.main_program.global_block()
+        ring_id = self.global_ring_id
+        record_idx, allreduce_input_vars, allreduce_output_vars = [], [], []
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py:213]: add some comment

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 11]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
...
+                        else:
+                            param = block.var(param_name)
+
+                    if len(allreduce_output_vars) == 0:
+                        allreduce_output_vars.append([grad])
+                        allreduce_input_vars.append([param])
+                        if self.allreduce_fusion_split_num == 1:
+                            record_idx.append([idx, idx])
+                            continue
+                        record_idx.append([0, idx])
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py]: append([-1, idx]) maybe better

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 12]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
...
+                            record_idx.append([idx, idx])
+                            continue
+                        if idx != len(list(enumerate(block.ops))) - 1:
+                            record_idx.append([0, idx])
+                    else:
+                        allreduce_output_vars[-1].append(grad)
+                        allreduce_input_vars[-1].append(param)
+                        record_idx[-1][0] = idx
+
+                if record_idx[-1][0] == 0:
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py]: == -1 maybe better

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 13]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
...
+                assert len(op_role_var) % 2 == 0
+                for i in range(0, len(op_role_var), 2):
+                    param_name = op_role_var[i]
+                    param = block.var(param_name)
+                    grad_name = op_role_var[i + 1]
+                    grad = block.var(grad_name)
+                    if param.is_distributed:
+                        continue
+                    if ".cast_fp16@GRAD" in grad_name:
+                        param_name = param_name + ".cast_fp16"
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py:231]: don't understand

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 14]
```diff
@@ -194,3 +204,249 @@ def _insert_allreduce_ops(self):
...
+                            name=unique_name.generate(
+                                'FusedOutput_{}_{}'.format(start, id_ +
+                                                           offset)),
+                            dtype=done_output_var[0].dtype,
+                            persistable=False,
+                            stop_gradient=True)
+                        self.vars['FusedOutput_{}_{}'.format(start, id_ +
+                                                             offset)] = tmp_var
+
+                        block._insert_op(
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py:291]: 不太对的亚子=。=

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 15]
```diff
@@ -194,3 +204,248 @@ def _insert_allreduce_ops(self):
...
+
+        self.vars = collections.OrderedDict()
+        index, offset_pos, pos, offset = 0, 0, 0, 0
+        start, end = record_idx[index]
+        men_list = [end, start]
+
+        for idx, op in reversed(list(enumerate(block.ops))):
+            if idx == start:
+                pos = 0
+                flag = True if end == men_list[-1] else False
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py:283]: 这个flag有啥作用

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 16]
```diff
@@ -194,3 +204,248 @@ def _insert_allreduce_ops(self):
...
+                                "dtype": done_output_var[0].dtype
+                            })
+                        pos += 1
+                offset_pos = pos
+
+                for id_ in range(len(done_output_vars)):
+                    if flag:
+                        block._insert_op(
+                            end + id_ + pos + 1,
+                            type='c_allreduce_sum',
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py:344]: need `c_sync_comm_stream` if use comm_stream

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 17]
```diff
@@ -194,3 +204,248 @@ def _insert_allreduce_ops(self):
...
+                                    start, id_ + offset)]
+                            },
+                            outputs={
+                                'Out': self.vars['FusedOutput_{}_{}'.format(
+                                    start, id_ + offset)]
+                            },
+                            attrs={
+                                'ring_id': ring_id,
+                                'use_calc_stream': True
+                                if self._comm_stream else False,
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py]: if comm_stream, use_calc_stream=**True** ?

File: python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py [hunk 18]
```diff
@@ -194,3 +204,248 @@ def _insert_allreduce_ops(self):
...
+                        type='c_sync_comm_stream',
+                        inputs={'X': block.create_var()},
+                        outputs={'Out': block.create_var()},
+                        attrs={
+                            'ring_id': ring_id,
+                            OP_ROLE_KEY: OpRole.Backward
+                        })
+                    break
+
+    def _split_fuction(self,
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/raw_program_optimizer.py:401]: I think need some comment

File: paddle/fluid/framework/distributed_strategy.proto (hunk 1 above)
Comment (Baibaifan) [reply paddle/fluid/framework/distributed_strategy.proto]: done

**Review by wanghuancoder — DISMISSED** [review]
LGTM for unittest.skip

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by phlrain — APPROVED** [review]
LGTM for check_graph = False

**Pull request merged**